Clarify schedule on which clients should check (#82)
aarongable/draft-acme-ari: draft-ietf-acme-ari.md

In particular, cron-based clients may find they need to increase their run frequency to check ARI more frequently. Those clients will need to store information about failures so that increasing their run frequency doesn't lead to retrying failures without proper backoff. Typical information stored should include: number of failures for a given order (defined by the set of names on the order), and time of the most recent failure.

> A RenewalInfo object in which the `end` timestamp equals or precedes the `start` timestamp is invalid.

Reviewer: If it's invalid, what is a client to do about it?

> ## Schedule for checking RenewalInfo objects

Reviewer 2: In this section, is it more correct to be talking about "RenewalInfo _objects_" or "the RenewalInfo _resource_"? I'm honestly unsure.

Author: The resource, good point.

Reviewer 2: ```suggestion
## Schedule for checking the RenewalInfo resource
```

> RenewalInfo objects must be fetched frequently so that clients learn about renewal quickly. But requests must not overwhelm the server. This protocol uses the Retry-After header to indicate to clients how often to retry. Note that in other HTTP applications, Retry-After often indicates the earliest time to retry a request. In this protocol, it indicates both the earliest time and a target time.

Reviewer 2: ```suggestion
RenewalInfo objects need to be fetched frequently enough that clients learn about renewal quickly, but without overwhelming the server. This protocol uses the Retry-After header to indicate to clients how often to retry. Note that in other HTTP applications, Retry-After often indicates the earliest time to retry a request. In this protocol, it indicates both the earliest time and a target time.
```

> Clients SHOULD fetch a certificate's RenewalInfo object immediately after issuance. After that, clients SHOULD fetch the certificate's RenewalInfo as soon as possible after the time indicated in the Retry-After header (backoff on errors takes priority, though). Clients SHOULD set reasonable limits on the value in the Retry-After header. For instance, values under 1 minute should be treated as if they were one minute and values over one day should be treated as if they were one day.

Reviewer: Did you mean servers? Clients don't set the header. Why should a client presume the server is wrong?

Reasonable seems arbitrary. Most servers will provably just implement the example. Just say they must do this maybe?

Reviewer 2: ```suggestion
Clients SHOULD fetch a certificate's RenewalInfo object immediately after issuance. After that, clients SHOULD fetch the certificate's RenewalInfo as soon as possible after the time indicated in the Retry-After header (backoff on errors takes priority, though). Clients SHOULD set reasonable limits on the value in the Retry-After header. For example, values under one minute could be treated as if they were one minute, and values over one day could be treated as if they were one day.
```

Reviewer: Ah, okay, so maybe:

> Clients SHOULD set reasonable limits on any value read from the Retry-After header.

> Clients SHOULD stop checking RenewalInfo after a certificate is expired. Clients SHOULD stop checking RenewalInfo after they consider a certificate to be replaced (for instance, after a new certificate for the same identifiers has been received and configured).

Reviewer: Not MUST? Allowing them to check ARI afterwards seems like a waste of resources.

Author: https://datatracker.ietf.org/doc/html/rfc2119#section-6

> [6](https://datatracker.ietf.org/doc/html/rfc2119#section-6). Guidance in the use of these Imperatives
>
>   Imperatives of the type defined in this memo must be used with care
>   and sparingly.  In particular, they MUST only be used where it is
>   actually required for interoperation or to limit behavior which has
>   potential for causing harm (e.g., limiting retransmisssions)  For
>   example, they must not be used to try to impose a particular method
>   on implementors where the method is not required for
>   interoperability.

This probably falls under a similar heading as "limiting retransmissions", so yeah I think it makes sense to call this a MUST.

Reviewer 2: ```suggestion
Clients MUST stop checking RenewalInfo after a certificate is expired. Clients MUST stop checking RenewalInfo after they consider a certificate to be replaced (for instance, after a new certificate for the same identifiers has been received and configured).
```

### Error handling

Temporary errors include, for instance:

  - Connection timeout
  - Request timeout
  - 5xx HTTP errors.

On receiving a temporary error, clients SHOULD do exponential backoff with a capped number of tries. If all tries are exhausted, clients SHOULD treat the request as a long-term error.

Long term errors include, for instance: 

  - Retry-After is invalid or not present
  - RenewalInfo object is invalid
  - DNS lookup failure
  - Connection refused
  - Non-5xx HTTP error

> On receiving a long term error, clients SHOULD perform the next RenewalInfo fetch as soon as possible after six hours have passed (or some other locally configured default).

Reviewer: Ah so maybe this answers my first question.

Good section imo.

Is a 500 really short term though? Usually those require changes to the server to fix.

And is 6 hours an example? The way it's worded makes it sound like it can be anything and doesn't really matter.

Reviewer 3: From the perspective of the client a 500 is likely to be resolved without its intervention , by the server getting repaired.

Retrying after six hours provides enough latitude that the herd won't be stampeding for tens of hours after a fault, which seems workable.

> ### Server choice of Retry-After

Reviewer 2: I would move this section up to be immediately after the first paragraph of the parent section. That way the logical flow is
1. Here's why retry-after is important
2. Here's how servers should set it
3. Here's how clients should respect it

rather than having servers be an afterthought. I think this will reduce questions like Matt's question above, questioning why clients should have reasonable limits when it seems like servers should also have those reasonable limits.

> Servers set the Retry-After header based on their requirements on how quickly to perform a revocation. For instance, a server that needs to revoke certificates within 24 hours of notification of a problem might choose to reserve twelve hours for investigation, six hours for clients to fetch RenewalInfo, and six hours for clients to perform a renewal. Setting a small value for Retry-After means that clients can respond more quickly, but also incurs more load on the server. Servers should estimate their expected load based on the number of clients, keeping in mind that third parties may also monitor RenewalInfo endpoints.

Reviewer: This is 👍
